### DISCUSSION TITLE
Janet 1.35.0

### DISCUSSION THREAD

**Discussion by bakpakin** [janet-lang/janet#1458] [General]
## 1.35.0 - 2024-06-15

- Add `:only` argument to `import` to allow for easier control over imported bindings.
- Add extra optional `env` argument to `eval` and `eval-string`.
- Allow naming function literals with a keyword. This allows better stacktraces for macros without
  accidentally adding new bindings.
- Add `bundle/` module for managing packages within Janet. This should replace the jpm packaging
  format eventually and is much simpler and amenable to more complicated builds.
- Add macros `ev/with-lock`, `ev/with-rlock`, and `ev/with-wlock` for using mutexes and rwlocks.
- Add `with-env`
- Add `*module-make-env*`dynamic binding
- Add buffer/format-at
- Add long form command line options for readable CLI usage
- Fix bug with `net/accept-loop` that would sometimes miss connections.
* Build janet with Cosmopolitan Libc by @ahgamut in https://github.com/janet-lang/janet/pull/1440
* Add ffi/defbind-alias by @wishdev in https://github.com/janet-lang/janet/pull/1445
* Add LoongArch64 support by @znley in https://github.com/janet-lang/janet/pull/1448
* Fix buffer push uint max by @pnelson in https://github.com/janet-lang/janet/pull/1457

## New Contributors
* @ahgamut made their first contribution in https://github.com/janet-lang/janet/pull/1440
* @znley made their first contribution in https://github.com/janet-lang/janet/pull/1448

**Full Changelog**: https://github.com/janet-lang/janet/compare/v1.34.0...v1.35.0

<hr /><em>This discussion was created from the release <a href='https://github.com/janet-lang/janet/releases/tag/v1.35.0'>Janet 1.35.0</a>.</em>

**Comment by CFiggers** [comment]
I really like the `:only` option for `(import)`. Allowing long-form CLI options is a great improvement as well. 👍 Thanks for the release!

**Reply by sogaiu** [reply]
I agree.

---

On a side note, has anyone figured out a way to use the `:only` feature without having to quote things?

What I got working was:

```janet
(import spork/fmt :only ['format])
```

Losing the quote gave me a compile error:

```
$ janet
Janet 1.35.0-0d9e9991 linux/x64/gcc - '(doc)' for help
repl:1:> (import spork/fmt :only [format])
repl:1:25: compile error: unknown symbol format
```

---

Side note to a side note (^^;

Cool [use of `invert` to make a "set"](https://github.com/janet-lang/janet/blob/7cd106a10c27bde21f8ece863cb07f7cd8d1ff09/src/boot/boot.janet#L3045) in `merge-module`'s source!

**Reply by CFiggers** [reply]
It looks like you can quote the whole tuple to avoid the compiler error:

```
$ janet
Janet 1.35.0-0d9e9991 linux/x64/gcc - '(doc)' for help
repl:1:> (import spork/path :only '[abspath join])
@{_ @{:value <cycle 0>} path/abspath @{:private true} path/join @{:private true}}
```

Still requires a quote, but at least you can use only one instead of needing to quote every symbol individually.

**Reply by sogaiu** [reply]
Thanks!

That's much better.

**Reply by sogaiu** [reply]
Looks like the [quoting may not be necessary](https://github.com/janet-lang/janet/commit/fda0a081f59c03d851658bf8f6d12371631f08f1) starting in 1.35.2.

**Comment by sivers** [comment]
Can bundle currently be used instead of jpm to install libraries like circlet and janet-pq that have .c files?

**Reply by sogaiu** [reply]
I don't think there is any network fetching, but I think [this sample file](https://github.com/janet-lang/spork/blob/bundle-hooks/bundle/init.janet) shows how if one applies bits in spork, compilation of `.c` is possible.

Perhaps known already, but more discussion is/was [here](https://github.com/janet-lang/janet/issues/1444).
